ql/src/java/org/apache/hadoop/hive/ql/optimizer/ParallelEdgeFixer.java

```diff
@@ -325,6 +331,11 @@ public static Optional<Set<String>> colMappingInverseKeys(ReduceSinkOperator rs)
     Map<String, String> ret = new HashMap<String, String>();
     Map<String, ExprNodeDesc> exprMap = rs.getColumnExprMap();
     Set<String> neededColumns = new HashSet<String>();
+
+    if (!rs.getSchema().getColumnNames().stream().allMatch(exprMap::containsKey)) {
```

Comment:
with that optimization, should we remove the `neededColumns` completely?

Reply:
Sure, I removed neededColumns and added comments to explain the reason for inversion failure.

Reply:
I forgot to push my local changes; please check the commit 27f2d73 for the requested changes.

Reply:
@ngsg, is there a way to reproduce the issue with a smaller dataset?

Reply:
The added qfile, sharedwork_semi_2.q, reproduces the issue and it does not require large data.

Reply:
👍 thanks

Reply:
it looks like `exprMap` might have duplicate columns {columnName} with different mappings {e.getKey()}, so I think we need to restore to prev version (revert #27f2d73)

Reply:
Thanks for pointing that out. I reverted it to the previous version.